### PR TITLE
add no-logical-or-in-switch-case rule

### LINKED ISSUE #5: Rule: Logical OR operators should not be used in switch cases
[RSPEC-3616](https://jira.sonarsource.com/browse/RSPEC-3616)

### PR DESCRIPTION
**Pull request by stas-vilchik**
Fixes #5

### CODE REVIEW

File: sonarts-core/src/rules/noLogicalOrInSwitchCaseRule.ts [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+          `currently this case clause only works for ${left}.`,
+      );
+    }
+
+    super.visitCaseClause(node);
+  }
+
+  private pickBarBarLiteralChain(expression: ts.Expression) {
+    let current = expression;
+    let elements = 0;
```
Comment (vilchik-elena) [inline sonarts-core/src/rules/noLogicalOrInSwitchCaseRule.ts]: @stas-vilchik Not the best name for counter

File: sonarts-core/src/rules/noLogicalOrInSwitchCaseRule.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    while (ts.isBinaryExpression(current) && ts.isLiteralExpression(current.right)) {
```
Comment (vilchik-elena) [inline sonarts-core/src/rules/noLogicalOrInSwitchCaseRule.ts]: @stas-vilchik I don't understand where we check the kind of binary expression

File: sonarts-core/tests/rules/noLogicalOrInSwitchCaseRule/noLogicalOrInSwitchCaseRule.lint.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+  switch (x) {
+    case 1:
+      return 1;
+    case 2 || 3:
+      // ^^^^^^ {{Explicitly specify 2 separate cases that fall through; currently this case clause only works for "2".}}
+      return 2;
+    case "a" || "b" || "c" || "d":
+      // ^^^^^^^^^^^^^^^^^^^^^^^^ {{Explicitly specify 4 separate cases that fall through; currently this case clause only works for "a".}}
+      return 3;
+    case bar() || baz(): // OK, not literals
```
Comment (vilchik-elena) [inline sonarts-core/tests/rules/noLogicalOrInSwitchCaseRule/noLogicalOrInSwitchCaseRule.lint.ts]: @stas-vilchik Could you try to raise issue here as well? I still think it's a confusing code, and we excluded it from JS due to the noise

File: sonarts-sq-plugin/sonar-typescript-plugin/src/main/resources/org/sonar/l10n/typescript/rules/typescript/S3616.json [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+  "type": "BUG",
+  "status": "ready",
+  "remediation": {
+    "func": "Constant\/Issue",
+    "constantCost": "5min"
+  },
+  "tags": [
+    
+  ],
+  "defaultSeverity": "Major"
```
Comment (vilchik-elena) [inline sonarts-sq-plugin/sonar-typescript-plugin/src/main/resources/org/sonar/l10n/typescript/rules/typescript/S3616.json]: @stas-vilchik regenerate with new rule-api

File: sonarts-core/src/rules/noLogicalOrInSwitchCaseRule.ts (hunk 2 above)
Comment (stas-vilchik) [reply sonarts-core/src/rules/noLogicalOrInSwitchCaseRule.ts]: 😱

**Pull request merged**